cloudsync/providers/file.py

```diff
@@ -0,0 +1,504 @@
+            fp = None
+    else:
+        fp = casedpath(full_path)  # canonicalizes path to an existing file
+
+    if not fp:
+        log.debug("canonicalize doesn't yet support missing parent folders %s", full_path)
+        return full_path
+
+    return fp
+
+
+@dataclass
+class CacheEnt:
+    mtime = 0.0
+    qhash = b''
+    fhash = b''
+
+
+class Watcher(watchdog_events.FileSystemEventHandler):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def on_any_event(self, event):
+        self.parent._on_any_event(event)                # pylint: disable=protected-access
+
+
+class FileSystemProvider(Provider):                     # pylint: disable=too-many-instance-attributes, too-many-public-methods
+    """
+    FileSystemProvider is a provider that uses the filesystem, and full paths as the storage location.
+    """
+    default_sleep = 0.01
+    name = "filesystem"
+    oid_is_path = True
+    case_sensitive = detect_case_sensitive()
+    _max_queue = 10000
+    _test_event_timeout = 1
+    _test_event_sleep = 0.001
+    _test_creds: typing.Dict[str, str] = {}
+    _test_namespace = os.path.join(tempfile.gettempdir(), os.urandom(16).hex())
+
+    def __init__(self, namespace="/"):
+        """Constructor for FileSystemProvider."""
+        self._namespace = namespace
+        self._cursor = 0
+        self._latest_cursor = 0
+        self._events: typing.Deque[Event] = collections.deque([])
+        self._evlock = threading.Lock()
+        self._evoffset = 0
+        self._event_window = 1000
+        self._cache_enabled = True
+        self._hash_cache: typing.Dict[str, CacheEnt] = {}
+        super().__init__()
+        self._test_creds = {"key": "val"}
+        self._observer = None
+
+    @property
+    def namespace(self):
+        return self._namespace
+
+    @namespace.setter
+    def namespace(self, path):
+        if self._namespace == path:
+            return
+        log.info("set namespace %s", path)
+        self._namespace = path
+
+        with self._api():
+            if not os.path.exists(self._namespace):
+                os.mkdir(self._namespace)
+
+            if self._observer:
+                self._observer.stop()
+
+            self._observer = watchdog_observer()
+            self._observer.schedule(Watcher(self), path, recursive=True)
+            self._observer.start()
+
+    @property
+    def namespace_id(self):
+        return self._fpath_to_oid(self._namespace)
+
+    @namespace_id.setter
+    def namespace_id(self, oid):
+        self.namespace = self._oid_to_fpath(oid)
+
+    def connect_impl(self, creds):
+        log.debug("connect mock prov creds : %s", creds)
+
+        if not creds:
+            raise ex.CloudTokenError()
+
+        self._api("connect_impl", creds)
+
+        if self.connection_id is None or self.connection_id == "invalid":
+            return os.urandom(16).hex()
+
+        return self.connection_id
+
+    def get_quota(self):
+        if not self.connected:
```

Comment:
linux I know supports actual quotas besides the limits set by the free space on the disk. I think windows has it, too. Probably not critical for the 1st release of this provider, but maybe TODO: comment and/or ticket to add this feature?

Reply:
ticket added